Added endpoint for annual_rainfall_summary()

lloyddewit opened the pull request.

lloyddewit commented:
@chrismclarke Please could you review and merge?
thanks

chrismclarke reviewed (changes requested):
Thanks @lloyddewit 

Trying to run locally I'm getting an error
![image](https://user-images.githubusercontent.com/10515065/225872453-b8e75246-fed6-4b3f-b045-69012b70c641.png)

I expect this is just because I do not have R installed, although that begs the question should I need it? 

I'm guessing the issue is coming from the `rpy2` package, so just want to check whether that is required in this repo? If so could you please update the main README to include in the prerequisites (ideally to include the version of R we agreed on previously). We will also need to update the cloud build scripts and api so that the docker machine has R installed also

lloyddewit commented:
@chrismclarke Thanks for testing and the feedback. 
When I changed this repo to call `annual_rainfall_summaries()` in `epicsa-python`, I found that it created a dependency in this repo for `rpy2`. Presumably this means that the api will require R.
I suggest we discuss in our meeting next week. I can then update the README.

chrismclarke commented:
Thanks @lloyddewit - yeah it makes sense that R will need to be installed for anyone using the repo as the epicsa_python dep executes r scripts.

I tried installing the R 4.2.3 and got through that step, but now receiving issues installing NumPy. I'm assuming this is also a dep coming from epicsa-python. Looking at the requirements on that repo I see you've put python 3.11+ required but this repo has 3.10+ mentioned... Was 3.11 an active decision? If so could we also update those docs here

![image](https://user-images.githubusercontent.com/10515065/225893516-5e932ea0-9e11-4248-9317-5c0815cd78e6.png)

lloyddewit commented:
@chrismclarke 
- I am using `numpy==1.24.2`. I just pushed a new `requirements_dev.txt` file to this PR (created with `pip freeze > requirements_dev.txt`).
- I only specified Python >= 3.11 because this is what I'm using. Is there a requirment to support 3.10? I agree that I should update the README.

We can discuss in our meeting shortly.

chrismclarke commented:
Just took a look back at the opencdms repo and remember we used an `install_packages.R` file to manage. I've updated documentation to include this step in be87bbe

@lloyddewit could you review that file to check if the same dependencies are required? I've also included the other epicsa/data and epicsa/wrap packages, which will need to be manually updated as required

chrismclarke commented:
Think I'm finally there on deps, added rtools install note to readme in 5dac218
Now I'm just stuck on the working dir issue you mentioned on the call, so can you please fix this to make it a relative path and then hopefully things should be working.

![image](https://user-images.githubusercontent.com/10515065/226355786-b77814d0-92d7-4b30-8829-1accf132bd08.png)

We'll need a separate PR to update the dockerfile to add support for all the additional infrastructure, I'll prep that separately to this

lloyddewit commented:
@chrismclarke fantastic, I'm very happy with the commits you made. I can't formally approve because I'm the PR author.
I will add the relative path.
I will also try and remove `cdms.products` and update the `install_packages.R` and `README.md` files. This should simplify the dependencies. It may also allow us to remove `RTools` (not guaranteed).
The requirement to manually update the git repo SHA's is understood and noted.

chrismclarke commented:
Sounds good, yeah the additional rtools and packages really ramps up the build time (previously 2 minutes and now more like 20+!). 

But no worries if not, I can always try add better caching so that we don't have to build everything if only the local code changes (create a base image that has most of R and Python dependencies preconfigured)

lloyddewit commented:
@chrismclarke 
The API should now work with the latest epicsa-python version (with relative paths). 
If you wish to test, you'll need to create a working folder structure. I added instructions for this to the `README`. I also found that I needed to uninstall the old `epicsa-python` before upgrading to the new SHA:
```
pip uninstall epicsa-python
pip install -r .\requirements_dev.txt
```
If you get a chance to review, then let me know how you get on, thanks!

chrismclarke reviewed (changes requested):
Thanks @lloyddewit 
I'll test fully shortly, however can I request that we automate the step of creating/validating folder structures? 

It will quickly become tedious to ensure the right folders are always manually created for local devs (and I'll need to write a script to populate on the CI anyway). So would be better if python/R includes some step to check if the folders exist, and create if not (depending on where the most logical place to do so would be)

chrismclarke commented:
@lloyddewit 
I've tried running locally again, even when manually configuring the folder structures and adding the service account I still can't get the annual_summaries endpoint to work. 

The first issue I face is with the input parameters to test. I was able to figure out country had to be either "zm" or "mw" and that summaries could be empty (but also a list of available summaries can be found in child packages, e.g. `seasonal_rainfall`).

I've updated the API in ... to add stronger type definitions for this to make it obvious

Then I had to figure out what station id to use, so looked through the storage bucket and figured out that the `.rds` files line up with station IDs. 

Even with this, selecting a couple stations at random I get the error 
![image](https://user-images.githubusercontent.com/10515065/226635209-7beb0016-8cd7-40a1-ab65-b0a5150a7ea5.png)

In addition if I try to run the operations twice in a row it seems like it cannot set the working directory (assume it's already in it so looking for `working_data/working_data`)
rpy2.rinterface_lib.embedded.RRuntimeError: Error in (function (dir)  : cannot change working directory

So before I spend any more time on this, can you please confirm that you have indeed tested yourself and things are working as expected? I.e. the endpoint you've added annual rainfall summaries works to generate a response as expected? That would make my job as a reviewer significantly easier

lloyddewit commented:
@chrismclarke 
Thank you, all good points, I will try and resolve today or tomorrow.
I confirm that I tested this myself and it worked as expected. 
I apologize for not providing the test data. I will do this in the future. I agree that if you're missing something like this, then it's best to ask me rather than spending time trying to deduce what's needed.

lloyddewit commented:
@chrismclarke 
I added functionality to auto create the working data folder structures (using an R function written by @dannyparsons). I also moved `service-account.json` up one level. I changed the default parameters to data that should give a valid result (country = "zm", station_id = "01122"). 
The first time I execute, it worked as expected (I sent you the expected response on Slack).
However the second and subsequent times I execute, I get an error. I assume this is something to do with the path specification. I will look into this tomorrow.
I suggest that you wait until I've fixed this before you test.
Thanks

![image](https://user-images.githubusercontent.com/57253949/227196534-0737a18e-55c8-4c21-b79e-0a2e0c31fced.png)

chrismclarke commented:
> I assume this is something to do with the path specification. I will look into this tomorrow

When working on projects in node/JS I typically use utility functions to specify paths relative to a known/fixed file (instead of current working directory - which can change at runtime).

E.g. Node has a utility `__dirname__` which gets the absolute path to the directory the file is in, and can be used to define things like the main project root dir (e.g. the file's parent folder, or perhaps a few folders higher depending on folder structures)

I think python has similar options - hopefully using this you can get the absolute path to the proposed working directory for the current user's file system
https://stackoverflow.com/questions/3430372/how-do-i-get-the-full-path-of-the-current-files-directory

lloyddewit commented:
@chrismclarke Thank you for the suggestions, I used these to improve my Python function.
In fact, this specific error was caused by something different (the data folder path and account key location were set relative to the current working directory but the current working directory was changed by the same R function so each call set the path one level lower).
This should now be fixed. You should be able to call the function repeatedly with valid parameters. I tested and it worked as expected. Please could you also test?
thanks!

chrismclarke commented:
Still getting issues. The first one you previously mentioned but I had forgotten that even doing a `pip install` doesn't replace the git repo package, you mentioned before needing to uninstall however I think there is a nicer solution of providing a `-U` flag with the command as per: https://stackoverflow.com/a/30562333

The second issue is coming from epicsawrap, but that is because I didn't realise updates had also been made to that package since last review and again requires update, but now I have working fine.

So I'll push one extra commit to update the readme for install and then I think finally good to go
